File: tools-manage/invocationlab-admin/src/main/java/org/javamaster/invocationlab/admin/util/ThreadLocalUtils.java
package org.javamaster.invocationlab.admin.util;

import org.springframework.core.NamedThreadLocal;

import java.util.HashMap;

/**
 * @author yudong
 * @date 2022/11/11
 */
public class ThreadLocalUtils {
    private static final ThreadLocal<HashMap<String, Object>> THREAD_LOCAL = new NamedThreadLocal<>("ThreadLocalUtil");
    public static final String OLD_GAV_VERSION = "oldGavVersion";

    public static <T> T get(String key) {
        return getOrDefault(key, null);
    }

    @SuppressWarnings("unchecked")
    public static <T> T getOrDefault(String key, Object defaultVal) {
        HashMap<String, Object> map = THREAD_LOCAL.get();
        if (map == null) {
            return (T) defaultVal;
        }
        Object obj = map.get(key);
        if (obj == null) {
            return (T) defaultVal;
        }
        return (T) obj;
    }

    public static void set(String key, Object obj) {
        HashMap<String, Object> map = THREAD_LOCAL.get();
        if (map == null) {
            map = new HashMap<>();
            THREAD_LOCAL.set(map);
        }
        map.put(key, obj);
    }

    public static void remove(String key) {
        HashMap<String, Object> map = THREAD_LOCAL.get();
        if (map == null) {
            return;
        }
        map.remove(key);
        if (map.isEmpty()) {
            THREAD_LOCAL.remove();
        }
    }

}

File: tools-manage/invocationlab-admin-21/src/main/kotlin/org/javamaster/invocationlab/admin/util/ThreadLocalUtils.kt
package org.javamaster.invocationlab.admin.util

import org.springframework.core.NamedThreadLocal

/**
 * @author yudong
 * @date 2022/11/11
 */
object ThreadLocalUtils {
    private val THREAD_LOCAL: ThreadLocal<HashMap<String, Any>> = NamedThreadLocal("ThreadLocalUtil")
    const val OLD_GAV_VERSION: String = "oldGavVersion"

    @JvmStatic
    fun <T> get(key: String): T? {
        return getOrDefault(key)
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T> getOrDefault(key: String): T? {
        val map = THREAD_LOCAL.get()
        val obj = map[key]
        return obj as T
    }

    @JvmStatic
    fun set(key: String, obj: Any) {
        var map = THREAD_LOCAL.get()
        if (map == null) {
            map = HashMap()
            THREAD_LOCAL.set(map)
        }
        map[key] = obj
    }

    @JvmStatic
    fun remove(key: String) {
        val map = THREAD_LOCAL.get() ?: return
        map.remove(key)
        if (map.isEmpty()) {
            THREAD_LOCAL.remove()
        }
    }
}
